transport_plugins/jlink/iotile_transport_jlink/jlink.py

```diff
@@ -163,166 +160,82 @@ def _try_connect(self, connection_string):
         """
-
         known_commands = {
-            'read_memory': JLinkControlThread.DEBUG_READ_MEMORY,
-            'write_memory': JLinkControlThread.DEBUG_WRITE_MEMORY,
-            'program_flash': JLinkControlThread.PROGRAM_FLASH
+            'read_memory': self._jlink_async.debug_read_memory,
+            'write_memory': self._jlink_async.debug_write_memory,
+            'program_flash': self._jlink_async.program_flash
         }
 
-        cmd_code = known_commands.get(cmd_name)
-        if cmd_code is None:
-            callback(conn_id, self.id, False, None, "Unsupported command: %s" % cmd_name)
+        self._ensure_connection(conn_id, True)
 
-        def _on_finished(_name, retval, exception):
-            if exception is not None:
-                callback(conn_id, self.id, False, None, str(exception))
-                return
+        func = known_commands.get(name)
+        if name is None:
+            raise ArgumentError("Unsupported command: %s" % name)
 
-            callback(conn_id, self.id, True, retval, None)
+        return await func(self._device_info, self._control_info, cmd_args)
 
-        self._control_thread.command(cmd_code, _on_finished, self._device_info, self._control_info, cmd_args, progress_callback)
 
-    def connect_async(self, connection_id, connection_string, callback):
+    async def connect(self, conn_id, connection_string):
         """Connect to a device by its connection_string
 
-        This function asynchronously connects to a device by its BLE address
-        passed in the connection_string parameter and calls callback when
-        finished.  Callback is called on either success or failure with the
-        signature:
-
-        callback(conection_id, adapter_id, success: bool, failure_reason: string or None)
-
-        Args:
-            connection_string (string): A unique connection string that identifies
-                which device to connect to, if many are possible.
-            connection_id (int): A unique integer set by the caller for
-                referring to this connection once created
-            callback (callable): A callback function called when the
-                connection has succeeded or failed
+        This function asynchronously connects to a device by device type.
+        See iotile_transport_jlink_devices.py
         """
-        self._try_connect(connection_string)
+        await self._try_connect(connection_string)
+        self._setup_connection(conn_id, connection_string)
 
-        def _on_finished(_name, control_info, exception):
-            if exception is not None:
-                callback(connection_id, self.id, False, str(exception))
-                return
+        try:
+            self._control_info = await self._jlink_async.verify_control_structure(self._device_info, self._control_info)
+        except HardwareError as e:
+            logger.exception("RPC, streaming and tracing won't work, but device still can be flashed or debugged")
```

Comment:
I would move this down to `warning` level and add `exc_info=True` to still attach the exception information to the event.  It's technically expected that this could happen in some situations so it's not a proper error,

Reply:
Will do so